content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+      @unroll sorted: Boolean = true,
+      nameMapper: String => Option[String] = Util.kebabCaseNameMapper
+  ): Either[String, T] = ???
+```
+
+This allows the developer to write the minimal amount of code they _want_ to write,
+and add a single annotation to allow binary compatibility to old versions. In this
+case, we annotated `sorted` with `@unroll`, which generates forwarders that make
+`def constructEither` binary compatible with older versions that have fewer parameters,
+up to a version before `sorted` was added. Any existing method `def`, `class`, or
+`case class` can be evolved in this way, by addition of `@unroll` the first time
+a default argument is added to their signature after its initial definition.
+
+### Unrolling `def`s
+
+Consider a library that is written as follows:
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1) = s + n + b + l
+}
+```
+
+If over time a new default parameter is added:
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1, b: Boolean = true) = s + n + b + l
+}
+```
+
+And another
+
+```scala
+object Unrolled{
+   def foo(s: String, n: Int = 1, b: Boolean = true, l: Long = 0) = s + n + b + l
+}
+```
+
+This is a source-compatible change, but not binary-compatible: JVM bytecode compiled against an 
+earlier version of the library would be expecting to call `def foo(String, Int)`, but will fail
+because the signature is now `def foo(String, Int, Boolean)` or `def foo(String, Int, Boolean, Long)`.
+On the JVM this will result in a `MethodNotFoundError` at runtime, a common experience for anyone
+who upgrading the versions of their dependencies. Similar concerns are present with Scala.js and
+Scala-Native, albeit the failure happens at link-time rather than run-time
+
+`@unroll` is an annotation that can be applied as follows, to the first "additional" default
+parameter that was added (in this case, `b: Boolean = true`)
+
+
+```scala
+import scala.annotation.unroll
+
+object Unrolled{
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = s + n + b + l
+}
+```
+
+The `@unroll` annotation takes `def foo` and generates synthetic forwarders for the purpose
+of maintaining binary compatibility for old callers who may be expecting the previous signature.
+These forwarders do nothing but forward the call to the current implementation, using the
+given default parameter values:
+
+```scala
+import scala.annotation.unroll
+
+object Unrolled{
+   def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = s + n + b + l
+
+   def foo(s: String, n: Int, b: Boolean) = foo(s, n, b, 0)
+   def foo(s: String, n: Int) = foo(s, n, true, 0)
+}
+```
+
+As a result, old callers who expect `def foo(String, Int, Boolean)` or `def foo(String, Int, Boolean, Long)`
+can continue to work, even as new parameters are added to `def foo`. The only restriction is that 
+new parameters can only be added on the right, and they must be provided with a default value.
```

Comment:
It would be good to add an explanation of what happens when there are multiple extensions of a method in different versions. Say in version 2 write 
```scala
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0)
```
and code like `foo("", 1, false, -1L)` compiles against that. Then in version 3  I would like another parameter with default to `foo` like this:
```scala
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0, f: Float = 0.0f)
```
How do I make the previous call `foo("", 1, false, -1L)` work? I assume I'll need another `@unroll`. What would the generated code be in this case?

Reply:
As currently written, `@unroll` generates all forwarders for parameters to the right of the annotation, and `@unrollOnly` generates a single forwarder for only that parameter.

We could also flip it around, and call them `@unrollAll` and `@unroll`, but the semantics would be the same. Just a question of naming

Reply:
So, assuming `unroll`, would I write
```scala
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0, @unroll f: Float = 0.0f)
```
? And what would that generate?

Reply:
I think a combination of something similar to `@unrollAll` + `@unroll` is perhaps most intuitive and with a lesser risk for misinterpretation. Or perhaps the latter should be called `@unrollSingle` to be explicit - I mean it is not easy to know what just `@unroll` means...

Reply:
Can we clarify what would happen under `@unroll` before discussing alternatives?

Reply:
As the proposal is currently written, it's

```scala
// Invalid, multiple `@unroll`s not allowed
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0, @unroll f: Float = 0.0f)

// Single `@unroll` generates forwarders for every method to the right
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0, f: Float = 0.0f)
// Generated
def foo(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0)
def foo(s: String, n: Int = 1, @unroll b: Boolean = true)
def foo(s: String, n: Int = 1)

// `@unrollOnly`s generates forwarders only for annotated parameters
def foo(s: String, n: Int = 1, @unrollOnly b: Boolean = true, l: Long = 0, @unrollOnly f: Float = 0.0f)
// Generated
def foo(s: String, n: Int = 1, @unrollOnly b: Boolean = true, l: Long = 0)
def foo(s: String, n: Int = 1)
```

We could swap out `@unroll`/`@unrollOnly` with `@unrollAll`/`@unroll` or `@unrollAll`/`@unrollOnly` or `@unroll.all`/`@unroll.only` if people think that's more intuitive. The naming is pretty arbitrary and I don't want to spend too much time discussing it up front

Reply:
Sorry for entering "naming mode" :) 
(I just wanted to give my initial feedback/thoughts...)